plugins/mntre/fu-mntre-device.c

```diff
@@ -0,0 +1,181 @@
+ */
+
+#include "config.h"
+
+#include <stdio.h>
+
+#include "fu-mntre-device.h"
+#include "mntre_reset.h"
+
+struct _FuMntreDevice {
+	FuUsbDevice parent_instance;
+};
+
+G_DEFINE_TYPE(FuMntreDevice, fu_mntre_device, FU_TYPE_USB_DEVICE)
+
+static gboolean
+fu_mntre_device_get_firmware_version(FuMntreDevice *self,
+				     GError **error,
+				     struct mntre_reset_firmware_version *firmware_version)
+{
+	g_autoptr(GError) error_local = NULL;
+	gsize actual_len = 0;
+
+	guint16 reset_interface = 2; // TODO
+
+	FuUsbDeviceClaimFlags flags = 0;
+
+	if (!fu_usb_device_claim_interface(FU_USB_DEVICE(self), reset_interface, flags, error)) {
+		g_prefix_error(error, "failed to claim HID interface: ");
+		return FALSE;
+	}
+
+	if (!fu_usb_device_control_transfer(FU_USB_DEVICE(self),
+					    FU_USB_DIRECTION_DEVICE_TO_HOST,
+					    FU_USB_REQUEST_TYPE_CLASS,
+					    FU_USB_RECIPIENT_INTERFACE,
+					    MNTRE_RESET_GET_FIRMWARE_VERSION,
+					    0,
+					    reset_interface,
+					    (guint8 *)firmware_version,
+					    sizeof(*firmware_version),
+					    &actual_len,
+					    2000,
+					    NULL,
+					    &error_local)) {
+		g_propagate_prefixed_error(error,
+					   g_steal_pointer(&error_local),
+					   "failed to read firmware version: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_mntre_device_reset_into_bootsel(FuMntreDevice *self, GError **error)
+{
+	g_autoptr(GError) error_local = NULL;
+
+	guint16 reset_interface = 2; // TODO
+
+	FuUsbDeviceClaimFlags flags = 0;
+
+	if (!fu_usb_device_claim_interface(FU_USB_DEVICE(self), reset_interface, flags, error)) {
+		g_prefix_error(error, "failed to claim HID interface: ");
+		return FALSE;
+	}
+
+	if (!fu_usb_device_control_transfer(FU_USB_DEVICE(self),
+					    FU_USB_DIRECTION_HOST_TO_DEVICE,
+					    FU_USB_REQUEST_TYPE_CLASS,
+					    FU_USB_RECIPIENT_INTERFACE,
+					    MNTRE_RESET_REQUEST_BOOTSEL,
+					    0,
+					    reset_interface,
+					    NULL,
+					    0,
+					    NULL,
+					    2000,
+					    NULL,
+					    &error_local)) {
+		if (g_error_matches(error_local, FWUPD_ERROR, FWUPD_ERROR_READ)) {
+			g_debug("ignoring expected error %s", error_local->message);
+		} else {
+			g_propagate_prefixed_error(error,
+						   g_steal_pointer(&error_local),
+						   "failed to restart device: ");
+			return FALSE;
+		}
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_mntre_device_detach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuMntreDevice *self = FU_MNTRE_DEVICE(device);
+
+	if (!fu_mntre_device_reset_into_bootsel(self, error))
+		return FALSE;
+
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static gboolean
+fu_mntre_device_ensure_version(FuMntreDevice *self, GError **error)
+{
+	g_autofree gchar *version = NULL;
+	struct mntre_reset_firmware_version firmware_version = {};
+
+	if (!fu_mntre_device_get_firmware_version(self, error, &firmware_version))
```

Comment:
ahh, rename this fu_mntre_device_ensure_firmware_version() and then call fu_device_set_version_raw() in that with the firmware_version.version

Reply:
I fixed this.